### PR TITLE
Jinja2 enhancements

### PR DESCRIPTION
**Pull request by TomekTrzeciak**
The future plan is to have proper API for Cylc, but this future is still quite far away. In the meantime, we could have some improvements to make the templating experience better. This PR adds two such improvements:
1) Support for Jinja2Tests and Jinja2Globals to complement the existing Jinja2Filters (thus enabling custom additions to all three namespaces used in Jinja2). This is useful as some Jinja2 functionality is awkward to use without them, e.g., `select`/`reject` filters, which use the `tests` namespace.
2) Support for importing Python modules in the same way as the template modules. This is useful, because the default Jinja2 functionality is fairly bare bones and there are some glaring omissions, like no Python `zip` equivalent, which makes some basic programming tasks quite awkward. Rather than keep adding useful bits of functionality one-by-one, it is left up to the user to decide which packages are of the most use.

TODO:
- [x] CUG documentation (I didn't want to spend time on it before hearing some feedback first)

### CODE REVIEW

**Comment by hjoliver** [comment]
This definitely sounds useful, no objections from me.  A couple of examples would be good, so I don't need to work it out (short of time!) - presumably your CUG documentation will add that.

**Comment by matthewrmshin** [comment]
Some test failures here. Should be easy enough to fix.

**Comment by TomekTrzeciak** [comment]
Added docs and rebased. @hjoliver, for an example see the new section "Importing additional Python modules" in the docs.

**Comment by matthewrmshin** [comment]
Codacy is hard to please, but it may have a point. Do you think this one is easy to solve?

**Comment by TomekTrzeciak** [comment]
I believe this PR is ready for review now.

File: lib/parsec/jinja2support.py [hunk 1]
```diff
@@ -24,12 +24,54 @@
...
+
+    # no source access for this loader
+    has_source_access = False
+
+    def __init__(self, prefix='__python__'):
+        self._templates = {}
+        # prefix that can be used to avoid name collisions with template files
+        self._python_namespace_prefix = prefix + '.'
+
+    def load(self, environment, name, glbs=None):
```
Comment (matthewrmshin) [inline lib/parsec/jinja2support.py]: Please add:
* A doc string.
* A comment on why we are good to ignore Codacy here.

**Comment by TomekTrzeciak** [comment]
Codacy drives me nuts :rage: I've fixed the complaints but this triggered hotspot detection due to frequent changes. Agrhhh...
Hopefully this will go away after the merge.

**Comment by hjoliver** [comment]
Codacy is useful, but yeah, it can be maddening.  We don't require a complete "pass" when it's obviously full of :hankey:

**Comment by matthewrmshin** [comment]
(Branch in small conflict.)

**Comment by TomekTrzeciak** [comment]
Fixed, should merge cleanly now.

**Pull request merged**